src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.Fixer.cs

```diff
@@ -0,0 +1,214 @@
+    /// - Stream.WriteAsync(Byte[], Int32, Int32, CancellationToken)
+    /// - Stream.ReadAsync(Byte[], Int32, Int32)
+    /// - Stream.ReadAsync(Byte[], Int32, Int32, CancellationToken)
+    ///
+    /// Preferred methods (available since .NET Standard 2.1 and .NET Core 2.1):
+    ///
+    /// - Stream.WriteAsync(ReadOnlyMemory{Byte}, CancellationToken)
+    /// - Stream.ReadAsync(Memory{Byte}, CancellationToken)
+    ///
+    /// </summary>
+    public abstract class PreferStreamAsyncMemoryOverloadsFixer : CodeFixProvider
+    {
+        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(PreferStreamAsyncMemoryOverloads.RuleId);
+
+        public sealed override FixAllProvider GetFixAllProvider() =>
+            WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Document doc = context.Document;
+            CancellationToken ct = context.CancellationToken;
+            SyntaxNode root = await doc.GetSyntaxRootAsync(ct).ConfigureAwait(false);
+
+            if (!(root.FindNode(context.Span, getInnermostNodeForTie: true) is SyntaxNode node))
+            {
+                return;
+            }
+
+            SemanticModel model = await doc.GetSemanticModelAsync(ct).ConfigureAwait(false);
+
+            if (!(model.GetOperation(node, ct) is IInvocationOperation invocation))
+            {
+                return;
+            }
+
+            // Defensive check to ensure the fix is only attempted on one of the 4 specific undesired overloads
+            if (invocation.Arguments.Length < 3 || invocation.Arguments.Length > 4)
+            {
+                return;
+            }
+
+            IArgumentOperation? bufferOperation = GetArgumentByPositionOrName(invocation.Arguments, 0, "buffer", out bool isBufferNamed);
+            if (bufferOperation == null)
+            {
+                return;
+            }
+
+            IArgumentOperation? offsetOperation = GetArgumentByPositionOrName(invocation.Arguments, 1, "offset", out bool isOffsetNamed);
+            if (offsetOperation == null)
+            {
+                return;
+            }
+
+            IArgumentOperation? countOperation = GetArgumentByPositionOrName(invocation.Arguments, 2, "count", out bool isCountNamed);
+            if (countOperation == null)
+            {
+                return;
+            }
+
+            // No nullcheck for this, because there is an overload that may not contain it
+            IArgumentOperation? cancellationTokenOperation = GetArgumentByPositionOrName(invocation.Arguments, 3, "cancellationToken", out bool isCancellationTokenNamed);
+
+            string title = MicrosoftNetCoreAnalyzersResources.PreferStreamAsyncMemoryOverloadsTitle;
+
+            Task<Document> fixInvocation = FixInvocation(doc, root, invocation, invocation.TargetMethod.Name,
```

Comment:
Invoking this method here will actually do the work of computing the fixed document, even before the user applied the fix. We should make it lazy so it is done only when user applies the code fix.
```suggestion
            Func<CancellationToken, Task<Document>> createChangedDocument = _ => FixInvocation(doc, root, invocation, invocation.TargetMethod.Name,
```